### PR TITLE
init fix for calculating producers #566

### PR DESCRIPTION
**Pull request by xmonader**
Initial fix for #566 

@zaibon  @ashraffouda  @rkhamis

### CODE REVIEW

File: lib/JumpScale/baselib/atyourservice81/Service.py [hunk 1]
```diff
@@ -148,36 +148,51 @@ def _initParent(self, actor, args):
...
+        # calculate the min required services and see if we should create new ones if auto is set
+        # create the services required till the minServices is reached.
+        # set add each to our producers and add ourself the their consumers list.
+        # maintain the parent relationship (parent is always a producer and we are always a consumer of the parent.)
 
-            if not j.data.types.list.check(instances):
-                instances = [instances]
+        for i, producer_model in enumerate(actor.model.dbobj.producers):
+            usersetservices = []
+            passedservicesnames = args.get(producer_model.argKey, [])
```
Comment (zaibon) [inline lib/JumpScale/baselib/atyourservice81/Service.py]: `args.get(producer_model.argKey, args.get(producer_role, ""))` was better cause we support the role and the actual attribute set in the schema.

in some schema the name of the attribute that define the producer is not the role of the producer.

File: lib/JumpScale/baselib/atyourservice81/Service.py [hunk 2]
```diff
@@ -148,36 +148,51 @@ def _initParent(self, actor, args):
...
+            if not j.data.types.list.check(passedservicesnames):
+                passedservicesnames = [passedservicesnames]
+            for x in passedservicesnames:
+                if x:
+                    sv = self.aysrepo.servicesFind(name=x, actor="%s.*" % producer_model.actorRole)
```
Comment (zaibon) [inline lib/JumpScale/baselib/atyourservice81/Service.py]: `actor=%s.*` match too much, see https://github.com/Jumpscale/jumpscale_core8/issues/592

File: lib/JumpScale/baselib/atyourservice81/Service.py [hunk 3]
```diff
@@ -148,36 +148,51 @@ def _initParent(self, actor, args):
...
+        # calculate the min required services and see if we should create new ones if auto is set
+        # create the services required till the minServices is reached.
+        # set add each to our producers and add ourself the their consumers list.
+        # maintain the parent relationship (parent is always a producer and we are always a consumer of the parent.)
 
-                    if producer_model.auto is False:
-                        raise j.exceptions.Input(message="could not find producer:%s for %s, found 0" %
-                                                 (producer_role, self), level=1, source="", tags="", msgpub="")
-                    else:
+        for i, producer_model in enumerate(actor.model.dbobj.producers):
```
Comment (rkhamis) [inline lib/JumpScale/baselib/atyourservice81/Service.py]: Why are you using `enumerate` here?

File: lib/JumpScale/baselib/atyourservice81/Service.py [hunk 4]
```diff
@@ -148,36 +148,51 @@ def _initParent(self, actor, args):
...
+            producer_role = producer_model.actorRole
+            usersetservices = []
+            passedservicesnames = args.get(producer_model.argKey, args.get(producer_role, ""))
+            if not j.data.types.list.check(passedservicesnames):
+                passedservicesnames = [passedservicesnames]
+            for x in passedservicesnames:
```
Comment (rkhamis) [inline lib/JumpScale/baselib/atyourservice81/Service.py]: could you please use more descriptive variable names? x is even used later, overwriting this value.

File: lib/JumpScale/baselib/atyourservice81/Service.py [hunk 5]
```diff
@@ -148,36 +148,51 @@ def _initParent(self, actor, args):
...
+                if x:
+                    sv = self.aysrepo.servicesFind(name=x, actor="%s(\..*)?" % producer_model.actorRole)
+                    usersetservices.extend(sv)
+
+            available_services = self.aysrepo.servicesFind(actor=producer_role)
+            available_services = list(set(available_services)-set(usersetservices))
+
+            extraservices = len(usersetservices) - producer_model.maxServices
+            if extraservices > 0:
+                raise j.exceptions.Input(message="There are more services than needed.", level=1, source="", tags="", msgpub="")
```
Comment (rkhamis) [inline lib/JumpScale/baselib/atyourservice81/Service.py]: error message is misleading
Should be something like "Specified services [usersetservices] are more than the maximum number (max) allowed"

File: lib/JumpScale/baselib/atyourservice81/Service.py [hunk 6]
```diff
@@ -148,36 +148,51 @@ def _initParent(self, actor, args):
...
-                        res.append(auto_actor.serviceCreate(instance="auto_%d" % instance, args={}))
-                elif len(res) > 1:
-                    raise j.exceptions.Input(message="could not find producer:%s for %s, found more than 1." %
-                                             (producer_role, self), level=1, source="", tags="", msgpub="")
-
-                producer_obj = res[0]
-                # add the service we consumer in the producers list
+                        available_services.append(auto_actor.serviceCreate(instance="auto_%s" % idx, args={}))
+                else:
+                    raise j.exceptions.Input(message="Auto isn't specified to auto create the required services.", level=1, source="", tags="", msgpub="")
```
Comment (rkhamis) [inline lib/JumpScale/baselib/atyourservice81/Service.py]: Misleading.
Something a long the lines of "Minimum number of services required is x and only y are provided. Auto is not set to create the missing services" would be a lot more helpful

**Pull request merged**